### ISSUE TITLE
Control which users can share the grain further and which do not

### ISSUE THREAD

**Issue by mitar** [sandstorm-io/sandstorm#1852]
It would be great to be able to define which users can share the grain further and which do not.

**Comment by neynah** [comment]
Could you tell us a bit more about the use case(s) where this would be beneficial for you?

**Comment by mitar** [comment]
I have a voting app/group decision making app. I would like that the user creates a new grain (they are admin) and then invite other users to vote on an issue (have permission to do so). But those other users should not be able to invite more people, because they might be outside of the allowed voters. (In my particular case I do not mind them inviting more people to watch the voting process, but they should not be able to allow that somebody else votes as well.)

**Comment by kentonv** [comment]
This is something we'd like to implement as a "policy": When you share, you could set the policy on the share that "this is not to be reshared".

Note that it is technically impossible to share with someone but prevent them from resharing maliciously -- they can always share their login credentials or implement some form of proxy. The "policy" would be what capability people call "voluntary oblivious compliance" -- it would adjust the UI so that the recipient has to go out of their way to reshare (e.g. by setting up such a proxy).

**Comment by kentonv** [comment]
I suppose for your voting app, what you really want is some way to grant exactly one vote to each user with whom you share. If they reshare, they are resharing their 1 vote, not allowing other people to add additional votes. This will become possible at some point.

**Comment by mitar** [comment]
> I suppose for your voting app, what you really want is some way to grant exactly one vote to each user with whom you share. If they reshare, they are resharing their 1 vote, not allowing other people to add additional votes. This will become possible at some point.

So if they give their login credentials, then yes, they are resharing their 1 vote. There can be some other ways to prevent that (requiring for example [ID cards with a hardware chip](https://en.wikipedia.org/wiki/Estonian_ID_card), so they would have to literally give to the other person their card).

But in my case app can make sure that each user can vote only once, but who can vote should (in my view) be handled by Sandstorm. So in this case Sandstorm preventing them from sharing would prevent them from allowing somebody else to vote.

**Comment by wrought** [comment]
@mitar @kentonv -- I'll point out the elephant in the room -- many people are accustomed to the sharing options available to google docs, something like:

- Who can view?
- Who can edit?
- Who controls sharing?

But really, it's just Unix, isn't it? 

- read (and execute, sorta)?
- write?
- owner (who controls sharing)?

Mitar wants:

- Maybe anyone (or specific persons with the link) can read the decisions history and comments
- Only certain invited users can write, or use the app to vote
- Only a certain subset of users with write ability can also share the ability to write, and therefore vote.

Imagine your company wants to use this app to vote -- you'd only want employees to vote, not their friends & relatives, right? But you may want to let friends and relatives enjoy the fact that your company is transparent about its decisions and discussions!

**Comment by mitar** [comment]
More that I think, I think each assigned permission should be in fact a tuple of two things (do I have permission, can I give it to others). Then, I could have users, which have permission, but cannot give it to others. Or users which even do not have a permission, but can give it to others (#1857 would be this in my case then, owner would have all permissions to give, but a limited ones to have themselves).

**Comment by kentonv** [comment]
Please be sure to read:

https://sandstorm.io/news/2015-05-05-delegation-is-the-cornerstone-of-civilization

Make sure to read the whole thing, because at the end we discuss the "please don't reshare" policy.

@wrought The model you are describing is the ACL model, whereas Sandstorm uses a capability model. ACL models often have the concept of "permission to edit the ACL", but this approach commonly fails in real-world situations. E.g. I may want to give you edit access to my doc and allow you to invite others to edit the doc, but I may not want you to be able to remove the other people I added. This seems like a perfectly reasonable desire, but the "permission to control permissions" approach simply can't express it. The capabilities + policies approach fits real-world use a lot more cleanly.

**Comment by mitar** [comment]
Yes, I like delegations as well. In fact my voting app is modeled around delegations for collaboration as well (see [these](http://mitar.tnode.com/post/73930434538/massive-online-collaborative-decision-making) [blog](http://mitar.tnode.com/post/73933896602/towards-a-better-group-decision-making) [posts](http://mitar.tnode.com/post/73983101095/peer-to-peer-voting-scheme), feel free to read them).

And I completely agree that the resharing is more of an UI thing that something else. I like the "please don’t reshare" message which user can then decide to obey or not.

Based on the blog and examples there, there is maybe a middle ground: you can pass capabilities around, but you loose them. So if I want somebody else to edit my document, I can pass capability to them, but then I cannot edit it, for the time being, until they return a token.

This could be an interesting middle ground because it would allow one to control who has access directly (without having to contact Alice), but on other hand would incentivize people to ask for capability back when they think the other person do not need it.

When we look at votes, this is pretty reasonable. You have a permission to vote. You can give somebody else this permission, to vote instead of you. But then you should not be able to vote anymore.

Alternatively, one other option would be that the app would be informed when capability was shared further by bypassing the policy. This could allow my voting app for example to take into account this (maybe giving each of the people 0.5 of the vote). And some other apps could maybe be able to evaluate if there is some leak happening.

**Comment by mitar** [comment]
Hm, maybe the issue is that my Meteor app sees only permissions, and not capabilities. If I could see directly capability tokens, and they would be unique, I could follow which of users have the same capability token and would be able to know in the app who and how people shared it. (Similar to what you are writing about auditing in your blog post.)

So maybe the point is, that there are two types of users:
* those who can just share their capability token further, and the same one is shared
* those who can issue a new capability token, so a new one is shared further

Admins could be the then the second one, and other users would be the first one. So instead of preventing sharing like ACL is doing and requiring that admin changes sharing settings, everyone can share. The difference is only that admin can create a token for somebody else (so that auditing can work from them on), while other users can just share their own token further.

This would work well with my app as well. I would just count how many different tokens are there, this is how many votes is possible. If somebody gave their token further, I could just split the vote among all for that token.

This could maybe tie nicely with #1857 as well. That owners would be able to issue new tokens, so the app would be able to differentiate between admin token (genesis token) and other tokens. And then adapt UI accordingly.
